core_impl/src/main/java/com/google/instrumentation/trace/TraceExporterImpl.java

```diff
@@ -0,0 +1,150 @@
+    workerThread.registerServiceHandler(name, serviceHandler);
+  }
+
+  @Override
+  public void unregisterServiceHandler(String name) {
+    workerThread.unregisterServiceHandler(name);
+  }
+
+  /**
+   * Adds a {@link SpanImpl} to be exported.
+   *
+   * @param span the {@code SpanImpl} that will be exported.
+   */
+  void addSpan(SpanImpl span) {
+    workerThread.addSpan(span);
+  }
+
+  private static final class WorkerThread extends Thread {
+    // Protects all variables.
+    private final Object monitor = new Object();
+
+    @GuardedBy("monitor")
+    private final List<SpanImpl> spans;
+
+    @GuardedBy("monitor")
+    private final Map<String, ServiceHandler> serviceHandlers =
+        new HashMap<String, ServiceHandler>();
+    // This must be immutable and changed only by swapping the reference, because reference
+    // store/load are atomic in Java.
+    private List<ServiceHandler> registeredServiceHandlers;
+    private final int bufferSize;
+    private final long scheduleDelay;
+
+    void addSpan(SpanImpl span) {
+      synchronized (monitor) {
+        this.spans.add(span);
+        if (spans.size() > bufferSize) {
+          monitor.notifyAll();
+        }
+      }
+    }
+
+    private void registerServiceHandler(String name, ServiceHandler serviceHandler) {
+      synchronized (monitor) {
+        serviceHandlers.put(name, serviceHandler);
+        registeredServiceHandlers = new ArrayList<ServiceHandler>(serviceHandlers.values());
+      }
+    }
+
+    private void unregisterServiceHandler(String name) {
+      synchronized (monitor) {
+        serviceHandlers.remove(name);
+      }
+    }
+
+    // Exports the list of SpanData to all the ServiceHandlers.
+    private void onBatchExport(List<SpanData> spanDataList) {
+      for (ServiceHandler serviceHandler : registeredServiceHandlers) {
+        serviceHandler.export(spanDataList);
+      }
+    }
+
+    private WorkerThread(int bufferSize, long scheduleDelay) {
+      spans = new LinkedList<SpanImpl>();
+      this.bufferSize = bufferSize;
+      this.scheduleDelay = scheduleDelay;
+      setDaemon(true);
+    }
+
+    // Returns an unmodifiable list of all buffered spans data.
+    private static List<SpanData> fromSpanImplToSpanData(List<SpanImpl> spans) {
+      List<SpanData> spanDatas = new ArrayList<SpanData>(spans.size());
+      for (SpanImpl span : spans) {
+        spanDatas.add(span.toSpanData());
+      }
+      return Collections.unmodifiableList(spanDatas);
+    }
+
+    @Override
+    public void run() {
+      while (true) {
+        List<SpanData> spanDataList;
+        synchronized (monitor) {
+          if (spans.size() < bufferSize) {
+            do {
+              try {
+                monitor.wait(scheduleDelay);
+              } catch (InterruptedException ie) {
+                // Preserve the interruption status as per guidance.
+                Thread.currentThread().interrupt();
```

Comment:
This is a daemon thread so it will not stop the shutdown process. But I agree with you probably is better to give up any work and return in case of an interruption.